У цьому домашньому завданні ми будемо працювати з набором даних Tips. Нижче інформація про нього.

**Про датасет:**

Tips - це датасет, який містить інформацію про рахунки та чайові в ресторані.

**Колонки датасету:**
- `total_bill` - загальна сума рахунку (долари)
- `tip` - розмір чайових (долари)
- `sex` - стать відвідувача (Male/Female)
- `smoker` - чи курить (Yes/No)
- `day` - день тижня (Thur/Fri/Sat/Sun)
- `time` - час відвідування (Lunch/Dinner)
- `size` - розмір групи (кількість людей)

**Джерело:** Оригінальні дані зібрані в одному з американських ресторанів у 1990-х роках. Скачати і переглянути набір даних можна тут: https://www.kaggle.com/datasets/sakshisatre/tips-dataset

А також ми вже вивантажили файл з даними і завантажили вам у Матеріали до цього модуля.




**1. Завантаження та огляд**

Завантажте датасет Tips (Чайові) з файла у матеріалах `tips.csv`.

Виведіть (кожен пункт оцінюється в 0.5 бала):
- Перші 5 рядків
- Розмір датасету (shape)
- Назви всіх колонок
- Інформацію про типи даних (info)

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('tips.csv')

In [3]:
print('Перші 5 рядків датасету tips')
df.head()

Перші 5 рядків датасету tips


total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [4]:
print(f'Розмір датасету tips: {df.shape}')

Розмір датасету tips: (244, 7)


In [5]:
print('Датасет містить наступні колонки')
df.columns

Датасет містить наступні колонки


Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

In [6]:
print('Інформація про тип даних')
df.info()

Інформація про тип даних
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   total_bill  244 non-null    float64
 1   tip         244 non-null    float64
 2   sex         244 non-null    object 
 3   smoker      244 non-null    object 
 4   day         244 non-null    object 
 5   time        244 non-null    object 
 6   size        244 non-null    int64  
dtypes: float64(2), int64(1), object(4)
memory usage: 13.5+ KB


Відповідно до інформації про тип даних всі числові метрики мають коректний тип даних, при цьому якісні параметри розпізнаються як object.
Датасет має 244 спостереження, при цьому пропуски в даних відсутні для всіх ознак.

**2. Базова вибірка даних**

Виберіть та виведіть (кожен пункт оцінюється в 0.5 бала):
- Колонку `total_bill` (загальний рахунок)
- Колонки `total_bill` та `tip` (чайові) разом
- Перші 10 рядків
- Рядок з індексом 5

In [7]:
print('Стовпець total_bill')
df['total_bill']

Стовпець total_bill


0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [8]:
print('Стовпці total_bill та tip')
df[['total_bill', 'tip']]

Стовпці total_bill та tip


total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

In [9]:
print('Перші 10 рядків датасету')
df.head(10)

Перші 10 рядків датасету


total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4
5       25.29  4.71    Male     No  Sun  Dinner     4
6        8.77  2.00    Male     No  Sun  Dinner     2
7       26.88  3.12    Male     No  Sun  Dinner     4
8       15.04  1.96    Male     No  Sun  Dinner     2
9       14.78  3.23    Male     No  Sun  Dinner     2

In [10]:
print('Рядок з індексом 5')
print(df.iloc[5])

Рядок з індексом 5
total_bill     25.29
tip             4.71
sex             Male
smoker            No
day              Sun
time          Dinner
size               4
Name: 5, dtype: object


**3. Фільтрація даних**

Відфільтруйте повний набір даних та виведіть (кожен пункт оцінюється в 0.5 бала):
- Всі рахунки більше 20 доларів (`total_bill > 20`)
- Всі відвідувачі-жінки (`sex == 'Female'`)
- Рахунки на вечерю (`time == 'Dinner'`)
- Рахунки на вечерю з сумою чеку більше 30 доларів

Для виведення результатів фільтрації:
- зберегти результат у окрему змінну
- показати кількість записів у відфільтрованому наборі
- вивести перші 5 записів відфільтрованого набору.

In [11]:
bill_20 = df[df['total_bill'] > 20]
print(f'Маємо {len(bill_20)} рахунків з сумою більше 20 доларів.')
print('Перші 5 записів:')
bill_20.head()


Маємо 97 рахунків з сумою більше 20 доларів.
Перші 5 записів:


total_bill   tip     sex smoker  day    time  size
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4
5       25.29  4.71    Male     No  Sun  Dinner     4
7       26.88  3.12    Male     No  Sun  Dinner     4

In [12]:
bill_female = df[df['sex'] == 'Female']
print(f'Маємо {len(bill_female)} рахунків від жінок')
print('Перші 5 записів:')
bill_female.head()

Маємо 87 рахунків від жінок
Перші 5 записів:


total_bill   tip     sex smoker  day    time  size
0        16.99  1.01  Female     No  Sun  Dinner     2
4        24.59  3.61  Female     No  Sun  Dinner     4
11       35.26  5.00  Female     No  Sun  Dinner     4
14       14.83  3.02  Female     No  Sun  Dinner     2
16       10.33  1.67  Female     No  Sun  Dinner     3

In [13]:
bill_dinner = df[df['time'] == 'Dinner']
print(f'Маємо {len(bill_dinner)} рахунків на вечерю')
print('Перші 5 записів:')
bill_dinner.head()


Маємо 176 рахунків на вечерю
Перші 5 записів:


total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [14]:
bill_dinner_30 = df[(df['time'] == 'Dinner') & (df['total_bill'] > 30)]
print(f'Маємо {len(bill_dinner_30)} рахунків на вечерю з сумою чеку більше 30 доларів')
print('Перші 5 записів:')
bill_dinner_30.head()

Маємо 27 рахунків на вечерю з сумою чеку більше 30 доларів
Перші 5 записів:


total_bill   tip     sex smoker  day    time  size
11       35.26  5.00  Female     No  Sun  Dinner     4
23       39.42  7.58    Male     No  Sat  Dinner     4
39       31.27  5.00    Male     No  Sat  Dinner     3
44       30.40  5.60    Male     No  Sun  Dinner     4
47       32.40  6.00    Male     No  Sun  Dinner     4

---

**4. Описова статистика**

Обчисліть (кожен пункт оцінюється в 0.5 бала):
- Середній розмір рахунку на всьому наборі даних
- Максимальні чайові
- Кількість пропущених значень у всіх колонках
- Скільки в наборі даних рахунків за обіди і за вечері

In [15]:
print(f'Середній розмір рахунку набору даних становить: {df['total_bill'].mean().round(2)}')

print(f'Максимальні чайові: {df['tip'].max()}')

Середній розмір рахунку набору даних становить: 19.79
Максимальні чайові: 10.0


In [16]:
print(f'Кількість пропущених значень у всіх стовпчиках')
df.isnull().sum()

Кількість пропущених значень у всіх стовпчиках


total_bill    0
tip           0
sex           0
smoker        0
day           0
time          0
size          0
dtype: int64

In [17]:
print('Кількість рахунків за часом')
df['time'].value_counts()

Кількість рахунків за часом


time
Dinner    176
Lunch      68
Name: count, dtype: int64

---

**5. Статистика із групуванням**

Обчисліть (кожен пункт оцінюється в 0.5 бала):
- Середній рахунок за статтю (groupby `sex`)
- Середні чайові за днем тижня (groupby `day`)
- Скільки відвідувачів у кожен день

In [18]:
print('Середній рахунок за статтю')
df.groupby('sex')['total_bill'].mean().round(2)

Середній рахунок за статтю


sex
Female    18.06
Male      20.74
Name: total_bill, dtype: float64

Загалом чоловіки мають вищий середній чек, ніж жінки.

In [19]:
print('Середні чайові за днем тижня')
df.groupby('day')['tip'].mean().round(2)

Середні чайові за днем тижня


day
Fri     2.73
Sat     2.99
Sun     3.26
Thur    2.77
Name: tip, dtype: float64

Найвищий середній чек спостерігається у неділю, при цьому в п'ятницю найменший.

In [20]:
print('Кількість відвідувачів по днях')
df.groupby('day')['total_bill'].count()

Кількість відвідувачів по днях


day
Fri     19
Sat     87
Sun     76
Thur    62
Name: total_bill, dtype: int64

Найменша кількість відвідувачів спостерігається у п'ятницю, при цьому, скоріш за все, вони замовляють їжу низької-середньої вартості, оскільки саме в даний день спостерігається найменший середній чек. При цьому можемо припустити, що найвищий середній чек у неділю забезпечується придбанням дорожчої продукції, оскільки найбільша кількість відвідувачів спостерігається в суботу.

---

**6. Створення нової колонки**

Створіть нову колонку `tip_percentage` - відсоток чайових від загальної суми рахунку.
Виведіть перші 10 рядків з новою колонкою.

In [21]:
df['tip_percentage'] = (df['tip'] / df['total_bill']).round(2)
df.head(10)

total_bill   tip     sex smoker  day    time  size  tip_percentage
0       16.99  1.01  Female     No  Sun  Dinner     2            0.06
1       10.34  1.66    Male     No  Sun  Dinner     3            0.16
2       21.01  3.50    Male     No  Sun  Dinner     3            0.17
3       23.68  3.31    Male     No  Sun  Dinner     2            0.14
4       24.59  3.61  Female     No  Sun  Dinner     4            0.15
5       25.29  4.71    Male     No  Sun  Dinner     4            0.19
6        8.77  2.00    Male     No  Sun  Dinner     2            0.23
7       26.88  3.12    Male     No  Sun  Dinner     4            0.12
8       15.04  1.96    Male     No  Sun  Dinner     2            0.13
9       14.78  3.23    Male     No  Sun  Dinner     2            0.22

---

**7. Сортування**

Відсортуйте датасет і виведіть результат сортування в кожному випадку окремо (кожен пункт оцінюється в 0.5 бала):
- За розміром рахунку (зростання)
- За чайовими (спадання)

Додатково виведіть топ-5 найбільших чайових використовуючи метод `nlargest`.

In [22]:
print('Відсортований датасет за розміром рахунку у порядку зростання')
df.sort_values('total_bill')

Відсортований датасет за розміром рахунку у порядку зростання


total_bill    tip     sex smoker   day    time  size  tip_percentage
67         3.07   1.00  Female    Yes   Sat  Dinner     1            0.33
92         5.75   1.00  Female    Yes   Fri  Dinner     2            0.17
111        7.25   1.00  Female     No   Sat  Dinner     1            0.14
172        7.25   5.15    Male    Yes   Sun  Dinner     2            0.71
149        7.51   2.00    Male     No  Thur   Lunch     2            0.27
..          ...    ...     ...    ...   ...     ...   ...             ...
182       45.35   3.50    Male    Yes   Sun  Dinner     3            0.08
156       48.17   5.00    Male     No   Sun  Dinner     6            0.10
59        48.27   6.73    Male     No   Sat  Dinner     4            0.14
212       48.33   9.00    Male     No   Sat  Dinner     4            0.19
170       50.81  10.00    Male    Yes   Sat  Dinner     3            0.20

[244 rows x 8 columns]

In [23]:
print('Відсортований датасет за чайовими у порядку спадання')
df.sort_values('tip', ascending = False)

Відсортований датасет за чайовими у порядку спадання


total_bill    tip     sex smoker   day    time  size  tip_percentage
170       50.81  10.00    Male    Yes   Sat  Dinner     3            0.20
212       48.33   9.00    Male     No   Sat  Dinner     4            0.19
23        39.42   7.58    Male     No   Sat  Dinner     4            0.19
59        48.27   6.73    Male     No   Sat  Dinner     4            0.14
141       34.30   6.70    Male     No  Thur   Lunch     6            0.20
..          ...    ...     ...    ...   ...     ...   ...             ...
0         16.99   1.01  Female     No   Sun  Dinner     2            0.06
111        7.25   1.00  Female     No   Sat  Dinner     1            0.14
92         5.75   1.00  Female    Yes   Fri  Dinner     2            0.17
67         3.07   1.00  Female    Yes   Sat  Dinner     1            0.33
236       12.60   1.00    Male    Yes   Sat  Dinner     2            0.08

[244 rows x 8 columns]

In [24]:
print('Топ-5 найбільших чайових')
df.nlargest(5, 'tip')

Топ-5 найбільших чайових


total_bill    tip   sex smoker   day    time  size  tip_percentage
170       50.81  10.00  Male    Yes   Sat  Dinner     3            0.20
212       48.33   9.00  Male     No   Sat  Dinner     4            0.19
23        39.42   7.58  Male     No   Sat  Dinner     4            0.19
59        48.27   6.73  Male     No   Sat  Dinner     4            0.14
141       34.30   6.70  Male     No  Thur   Lunch     6            0.20